File: protocol-generator/elixir/lib/entity_core/base58.ex
defmodule EntityCore.Base58 do
  @moduledoc """
  Base58 (Bitcoin alphabet) encode/decode, hand-rolled (no Hex dependency).

  Used for peer-id formatting/parsing (V7 §1.2). Leading zero bytes map to a
  leading `"1"` each, per the standard Base58 convention.
  """

  @alphabet "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

  @doc "The Bitcoin Base58 alphabet (peer-id segment-character membership test)."
  @spec alphabet() :: binary()
  def alphabet, do: @alphabet

  @doc "Encode a binary as a Base58 string."
  @spec encode(binary()) :: binary()
  def encode(bin) when is_binary(bin) do
    zeros = leading_zeros(bin, 0)
    body =
      case :binary.decode_unsigned(bin, :big) do
        0 -> []
        n -> encode_int(n, [])
      end

    String.duplicate("1", zeros) <> IO.iodata_to_binary(body)
  end

  @doc """
  Decode a Base58 string back to a binary.

  Returns `{:ok, binary}` or `{:error, reason}`.
  """
  @spec decode(binary()) :: {:ok, binary()} | {:error, atom()}
  def decode(str) when is_binary(str) do
    chars = :binary.bin_to_list(str)

    if Enum.all?(chars, &(:binary.match(@alphabet, <<&1>>) != :nomatch)) do
      ones = leading_ones(chars, 0)
      n = Enum.reduce(chars, 0, fn c, acc -> acc * 58 + index_of(c) end)
      body = if n == 0, do: <<>>, else: :binary.encode_unsigned(n, :big)
      {:ok, :binary.copy(<<0>>, ones) <> body}
    else
      {:error, :invalid_base58}
    end
  end

  defp encode_int(0, acc), do: acc
  defp encode_int(n, acc), do: encode_int(div(n, 58), [:binary.at(@alphabet, rem(n, 58)) | acc])

  defp leading_zeros(<<0, rest::binary>>, n), do: leading_zeros(rest, n + 1)
  defp leading_zeros(_, n), do: n

  defp leading_ones([?1 | rest], n), do: leading_ones(rest, n + 1)
  defp leading_ones(_, n), do: n

  defp index_of(char), do: index_of(char, 0)
  defp index_of(char, i), do: if(:binary.at(@alphabet, i) == char, do: i, else: index_of(char, i + 1))
end

File: protocol-generator/ruby/lib/entity_core/base58.rb
# frozen_string_literal: true

require_relative "error"

module EntityCore
  # Base58 (Bitcoin alphabet) encode/decode, hand-rolled (no gem — dodges a dep
  # + a pin). Used for peer-id formatting/parsing (V7 §1.5). Each leading
  # zero byte maps to a leading "1", per the standard Base58 convention.
  module Base58
    ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
    INDEX = ALPHABET.each_char.with_index.to_h.freeze

    module_function

    # Encode an Encoding::BINARY String to a Base58 (UTF-8) String.
    def encode(bin)
      bytes = bin.b
      zeros = 0
      zeros += 1 while zeros < bytes.bytesize && bytes.getbyte(zeros).zero?

      n = bytes.empty? ? 0 : bytes.unpack1("H*").to_i(16)
      body = +""
      while n.positive?
        n, rem = n.divmod(58)
        body.prepend(ALPHABET[rem])
      end

      ("1" * zeros) + body
    end

    # Decode a Base58 String back to an Encoding::BINARY String. Raises
    # CodecError on a non-alphabet character.
    def decode(str)
      ones = 0
      ones += 1 while ones < str.length && str[ones] == "1"

      n = 0
      str.each_char do |c|
        idx = INDEX[c]
        raise CodecError, "invalid base58 character: #{c.inspect}" if idx.nil?

        n = (n * 58) + idx
      end

      body =
        if n.zero?
          String.new(encoding: Encoding::BINARY)
        else
          hex = n.to_s(16)
          hex = "0#{hex}" if hex.length.odd?
          [hex].pack("H*")
        end

      (("\x00".b * ones) + body)
    end
  end
end
